src/api/object.js

```diff
@@ -1,33 +1,259 @@
-      },
-      setData: (root, data, cb) => {
-        return send('object/patch/set-data', [root], null, data, cb)
-      },
-      appendData: (root, data, cb) => {
-        return send('object/patch/append-data', [root], null, data, cb)
-      },
-      addLink: (root, name, ref, cb) => {
-        return send('object/patch/add-link', [root, name, ref], null, null, cb)
+      addLink: promisify((multihash, dLink, options, callback) => {
+        if (typeof options === 'function') {
+          callback = options
+          options = {}
+        }
+        if (!options) {
+          options = {}
+        }
+        multihash = cleanMultihash(multihash, options)
+
+        send('object/patch/add-link', [multihash, dLink.name, bs58.encode(dLink.hash).toString()], null, null, (err, result) => {
+          if (err) {
+            return callback(err)
+          }
+          api.get(result.Hash, { enc: 'base58' }, callback)
+        })
+      }),
+      rmLink: promisify((multihash, dLink, options, callback) => {
+        if (typeof options === 'function') {
+          callback = options
+          options = {}
+        }
+        if (!options) {
+          options = {}
+        }
+        multihash = cleanMultihash(multihash, options)
+
+        send('object/patch/rm-link', [multihash, dLink.name], null, null, (err, result) => {
+          if (err) {
+            return callback(err)
+          }
+          api.get(result.Hash, { enc: 'base58' }, callback)
+        })
+      }),
+      setData: promisify((multihash, data, options, callback) => {
+        if (typeof options === 'function') {
+          callback = options
+          options = {}
+        }
+        if (!options) {
+          options = {}
+        }
+        multihash = cleanMultihash(multihash, options)
+
+        send('object/patch/set-data', [multihash], null, data, (err, result) => {
+          if (err) {
+            return callback(err)
+          }
+          api.get(result.Hash, { enc: 'base58' }, callback)
+        })
+      }),
+      appendData: promisify((multihash, data, options, callback) => {
+        if (typeof options === 'function') {
+          callback = options
+          options = {}
+        }
+        if (!options) {
+          options = {}
+        }
+        multihash = cleanMultihash(multihash, options)
+
+        send('object/patch/append-data', [multihash], null, data, (err, result) => {
+          if (err) {
+            return callback(err)
+          }
+          api.get(result.Hash, { enc: 'base58' }, callback)
+        })
+      })
+    }
+  }
+  return api
+}
+
+function cleanMultihash (multihash, options) {
+  if (Buffer.isBuffer(multihash)) {
+    if (options.enc) {
+      switch (options.enc) {
+        case 'base58': multihash = multihash.toString()
+      }
+    } else {
+      multihash = bs58.encode(multihash).toString()
+    }
+  } else if (typeof multihash === 'string') {
+    if (options.enc) {
+      switch (options.enc) {
```

Comment:
so that when someone add another encoding, they just add one clause instead of c&p a if clause